[CIS-789,CIS-790] Use custom keys for image caching and Load image thumbnails correctly

skorepak opened the pull request:
## In this PR
- I created image CDN protocol to filter out query parameters from image URL and add thumbnail size parameters to the URL
- used the provider to fix an issue with constantly reloading images 

## Links
https://stream-io.atlassian.net/browse/CIS-789
https://stream-io.atlassian.net/browse/CIS-790

VojtaStavik commented on Sources/StreamChatUI/Utils/ImageCDN.swift:
```diff
@@ -0,0 +1,39 @@
+
+import Foundation
+
+public var cachingKeyProvider: CachingKeyProvider = StreamCachingKeyProvider()
```
I'd love to see how far can we get if we used this inside `Compomnents`

skorepak replied:
Working on it 👍🏻

skorepak replied:
Tried multiple ways but I'm failing to pull it out of components without ExtraData on `UIImageView` the only way I can see right now is to subclass 
```
class StreamImageView<ExtraData: ExtraDataTypes>: UIImageVIew {}
```
and add `StreamImageView` to components ass well

skorepak replied:
I will open another PR with implementing the `UIImageView` subclass so we can decide if we want it or not. this PR can come trough

VojtaStavik commented on Sources/StreamChatUI/Utils/ImageCDN.swift:
```diff
@@ -0,0 +1,39 @@
+///
+/// The URL can contain extra parameters that are changing every time. Those parameters are effectively blocking the caching.
+/// By providing customised (filtered) key, the cache will work as expected.
+public protocol CachingKeyProvider {
```
Maybe `ImageCachingKeyProvider`?

VojtaStavik commented on Sources/StreamChatUI/Utils/ImageCDN.swift:
```diff
@@ -0,0 +1,39 @@
+    func cachingKey(forImage url: URL) -> String
+}
+
+open class StreamCachingKeyProvider: CachingKeyProvider {
+    public static var streamCDNURL = "stream-io-cdn.com"
+    
+    public func cachingKey(forImage url: URL) -> String {
+        let key = url.absoluteString
+        
+        guard
+            var components = URLComponents(url: url, resolvingAgainstBaseURL: true),
+            let host = components.host
+        else { return key }
+        
+        // Stream CDN
+        if host.contains(StreamCachingKeyProvider.streamCDNURL) {
+            components.query = nil
+            return components.string ?? key
+        }
+        
+        return key
+    }
+}
```
This probably doesn't have to be `open` as it's more an implementation detail of the default CDN.

VojtaStavik reviewed (commented):
Looks like the correct approach 👍

DominikBucher12 commented on Sources/StreamChatUI/Utils/ImageCDN.swift:
```diff
@@ -0,0 +1,37 @@
+}
+
+open class StreamCachingKeyProvider: CachingKeyProvider {
+    public static var streamCDNURL = "stream-io-cdn.com"
```
nit: `public static let`?? Why is it not constant?

evsaev commented on Sources/StreamChatUI/Utils/ImageCDN.swift:
```diff
@@ -0,0 +1,39 @@
+    func cachingKey(forImage url: URL) -> String
+}
+
+open class StreamCachingKeyProvider: CachingKeyProvider {
+    public static var streamCDNURL = "stream-io-cdn.com"
+    
+    public func cachingKey(forImage url: URL) -> String {
```
I'd step out from inheritance and make the `StreamCachingKeyProvider` a struct. We have a protocol and customers will build their own implementation of `CachingKeyProvider` from scratch anyway

skorepak replied:
You are right, good catch!

skorepak commented:
Added function to load thumbnail of preferred size and renamed protocol to ImageCDN. 
I'm wondering whether to limit the thumbnail URL just for our Image CDN. Anybody nows?

b-onc commented on Sources/StreamChatUI/Utils/ImageCDN.swift:
```diff
@@ -0,0 +1,74 @@
+///
+/// The URL can contain extra parameters that are changing every time. Those parameters are effectively blocking the caching.
+/// By providing customised (filtered) key, the cache will work as expected.
+public protocol ImageCDN {
```
This works more like a URL provider than CDN, so I suggest renaming this to `ImageURLProvider` or something like it, basically removing `CDN` from the name

skorepak replied:
The problem is that it's not providing URL, is providing URL modifiers/processors, so I was struggling with the name a little. I used the one suggested by @VojtaStavik in the task

VojtaStavik replied:
For sure `ImageCDN` is not the best name but I think it kinda works 🤷‍♂️ You'd need to implement these methods if you use your own CDN for images. We'll probably add more methods here as we add new features.

b-onc commented on Sources/StreamChatUI/Utils/ImageCDN.swift:
```diff
@@ -0,0 +1,74 @@
+    ///   - preferedSize: The requested thumbnail size.
+    ///   - crop: Crop type to be used for the thumbnail.
+    ///   - resize: Resize type to be used for the thumbnail.
+    func thumbnailURL(originalURL: URL, preferedSize: CGSize, crop: ImageCrop, resize: ImageResize) -> URL
```
This func is not actually being used?

skorepak replied:
No, not yet. I’m trying to figure out how/when to use it. We don't want to use it every time, but only when the thumbnail is required. Which I don't know yet.

skorepak replied:
It's being used now

b-onc commented on Sources/StreamChatUI/Utils/ImageCDN.swift:
```diff
@@ -0,0 +1,74 @@
+
+import UIKit
+
+public enum ImageCrop: String {
+    case top, bottom, left, right, center
+}
+
+public enum ImageResize: String {
+    case crop, scale, fill
+}
```
These should not public since they're params for Stream CDN. Custom CDN's can have their own params

skorepak replied:
It needs to be public because it's used in public protocol.

VojtaStavik replied:
Hmm, do you think we need to expose this to users? I'd be ok with hardcoding this to `fill + center` 🤔

VojtaStavik commented on Sources/StreamChatUI/Components.swift:
```diff
@@ -70,6 +70,9 @@ public struct _Components<ExtraData: ExtraDataTypes> {
     
     /// The view that shows a loading indicator.
     public var loadingIndicator: ChatLoadingIndicator.Type = ChatLoadingIndicator.self
+    
+    /// Object providing custom key to use for image cache.
+    public var imageCDN: ImageCDN = StreamImageCDN()
```
outdated description

VojtaStavik commented on Sources/StreamChatUI/Utils/ImageCDN.swift:
```diff
@@ -0,0 +1,74 @@
+    }
+    
+    public func thumbnailURL(originalURL: URL, preferedSize: CGSize, crop: ImageCrop, resize: ImageResize) -> URL {
+        guard var components = URLComponents(url: originalURL, resolvingAgainstBaseURL: true) else { return originalURL }
```
Here we also need to check that the URL is a Stream CDN URL

skorepak replied:
Yes, already fixed that locally 👍🏻

VojtaStavik commented on Sources/StreamChatUI/Utils/UIImageView+Extensions.swift:
```diff
@@ -3,14 +3,18 @@
 
 extension UIImageView {
     @discardableResult
-    func loadImage(
+    func loadImage<ExtraData: ExtraDataTypes>(
```
I think it's about time to add documentation to this method :)

VojtaStavik commented on Sources/StreamChatUI/Utils/UIImageView+Extensions.swift:
```diff
@@ -3,14 +3,18 @@
         from url: URL?,
         placeholder: UIImage? = nil,
         resizeAutomatically: Bool = true,
+        requestThumbnail: Bool = false,
```
TBH this API now feels a bit complex. In the end, it all comes to the question - do you need the image in a specific size? If yes, you should try to get the correct URL and get it.

Maybe there's a way to simplify this?

skorepak replied:
I was thinking about some enum with associated value like:
```
enum Resize {
    .resizeAfterDownload
    .requestThumbnail(size: CGSize?)
    .none
}
```
But you didn't like adding an extra option for ImageCrop and this would be the same. Adding extra public enum just for this. I would say it's a bit extensive, but not that complex. 

I'm open to suggestions on how to simplify this

VojtaStavik reviewed (approved):
LGTM, just a few comments to the API design 👍

skorepak commented on Sources/StreamChatUI/Utils/ImageCDN_Tests.swift:
```diff
@@ -0,0 +1,100 @@
+    func test_thumbnail_validStreamUrl_withParameters() {
+        let provider = StreamImageCDN()
+        
+        let url = URL(string: "https://wwww.stream-io-cdn.com/image.jpg?name=Luke")!
+        let thumbnailUrl =
+            URL(string: "https://wwww.stream-io-cdn.com/image.jpg?name=Luke&w=120&h=120&crop=center&resize=fill&ro=0")!
+        let processedURL = provider.thumbnailURL(
+            originalURL: url,
+            preferredSize: CGSize(width: 40, height: 40)
+        )
```
That's true! thanks for noticing 👍🏻

VojtaStavik commented on Sources/StreamChatUI/Utils/UIImageView+Extensions.swift:
```diff
@@ -79,3 +79,36 @@ private extension UIImageView {
+            return size
+        }
+
+        private let sizeProvider: () -> CGSize
```
Maybe this should return `CGSize?` rather than using `.zero` as the fallback?

VojtaStavik commented on Sources/StreamChatUI/Utils/UIImageView+Extensions.swift:
```diff
@@ -79,3 +79,36 @@ private extension UIImageView {
+        }
+
+        public var identifier: String {
+            "com.github.kean/nuke/lateResize"
```
Shouldn't we make this unique every time we create a processor?

skorepak replied:
I was trying to figure out if it matters or not. Nuke docs don't say anything and the original `Resize` is adding just the parameters of the processor. If you create two `Resize` processors with the same size parameter, they will have the same identifier. Since we have size closure and getting the size is waiting synchronously for the main queue (may be expensive), I decided to remove it.
